From e0bce01a56f6481ba34d08adc3d90444a92f7403 Mon Sep 17 00:00:00 2001
From: Jack Tan <jack.tan@windriver.com>
Date: Tue, 16 Dec 2008 19:06:57 +0800
Subject: [PATCH] rmi xlr clock event and clock source support

There are 8 timers in XLR's on-chip Programmable Interrupt Controller.
vCPU0 (the first boot up thread) use PIC's timer7 as its timer interrupt
source (clock event) and other 30 vCPUs use mips compare/count as its timer
interrupt source.

And we use PIC's timer6 as its hardware clock source.

Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 arch/mips/kernel/csrc-phoenix.c |   68 +++++++++++++++++++++++++++++++++++++++
 arch/mips/rmi/phoenix/time.c    |   37 +++++++++++++++++++++
 2 files changed, 105 insertions(+), 0 deletions(-)
 create mode 100644 arch/mips/kernel/csrc-phoenix.c
 create mode 100644 arch/mips/rmi/phoenix/time.c

diff --git a/arch/mips/kernel/csrc-phoenix.c b/arch/mips/kernel/csrc-phoenix.c
new file mode 100644
index 0000000..d848eba
--- /dev/null
+++ b/arch/mips/kernel/csrc-phoenix.c
@@ -0,0 +1,68 @@
+/*
+ *  Copyright 2008 Wind River Systems, Inc. 
+ *
+ *  Author: Jack Tan <jack.tan@windriver.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include <linux/clocksource.h>
+
+#include <asm/irq.h>
+#include <asm/ptrace.h>
+#include <asm/addrspace.h>
+#include <asm/time.h>
+#include <asm/cpu.h>
+#include <asm/cpu-features.h>
+
+#include <asm/rmi/pic.h>
+
+#define	PIC_TIMER6_MAX_VAL	0xffffffff
+
+cycle_t xlr_hpt_read(void)
+{
+	phoenix_reg_t *mmio = phoenix_io_mmio(PHOENIX_IO_PIC_OFFSET);
+	uint32_t counter = 0;
+
+	counter = phoenix_read_reg(mmio, PIC_TIMER_6_COUNTER_0);
+
+	return (cycle_t) (PIC_TIMER6_MAX_VAL - counter);
+}
+
+struct clocksource phoenix_clocksource = {
+	.name	= "phoenix-timer6",
+	.rating	= 200 + PIC_CLKS_PER_SEC / 10000000,
+	.read	= xlr_hpt_read,
+	.mask	= CLOCKSOURCE_MASK(32),
+	.flags	= CLOCK_SOURCE_IS_CONTINUOUS,
+};
+
+void __init phoenix_clocksource_init(void)
+{
+	struct clocksource *cs = &phoenix_clocksource;
+	phoenix_reg_t *mmio = phoenix_io_mmio(PHOENIX_IO_PIC_OFFSET);
+
+	/* Use PIC Timer 6 as a free running counter */
+	phoenix_write_reg(mmio, PIC_TIMER_6_MAXVAL_0, 0xffffffff);
+	phoenix_write_reg(mmio, PIC_TIMER_6_MAXVAL_1, 0x0);
+	/* we Don't need interrupts */
+	phoenix_write_reg(mmio, PIC_IRT_0_TIMER_6, 0);
+	phoenix_write_reg(mmio, PIC_IRT_1_TIMER_6,
+					  (1 << 31) | (0 << 30) | (1 << 6) | (PIC_TIMER_6_IRQ));
+	pic_update_control(1 << (8 + 6));
+
+	clocksource_set_clock(cs, PIC_CLKS_PER_SEC);
+	clocksource_register(cs);
+}
diff --git a/arch/mips/rmi/phoenix/time.c b/arch/mips/rmi/phoenix/time.c
new file mode 100644
index 0000000..470625d
--- /dev/null
+++ b/arch/mips/rmi/phoenix/time.c
@@ -0,0 +1,37 @@
+/*
+ *  Copyright 2008 Wind River Systems, Inc. 
+ *
+ *  Author: Jack Tan <jack.tan@windriver.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <linux/init.h>
+#include <asm/time.h>
+#include <asm/rmi/pic.h>
+#include <asm/rmi/sim.h>
+#include <asm/rmi/debug.h>
+
+extern void __init phoenix_clocksource_init(void);
+
+extern struct psb_info *prom_info;
+
+void __init plat_time_init(void)
+{
+	mips_hpt_frequency = (unsigned int)prom_info->cpu_frequency;
+
+	phoenix_clocksource_init();
+}
+
-- 
1.6.0.4

